gramine_ratls_attest.go

```diff
@@ -0,0 +1,99 @@
+package gramine_ratls
+
+// #cgo LDFLAGS: -ldl
+// #include <assert.h>
+// #include <stdlib.h>
+// #include <dlfcn.h>
+// #include <stdbool.h>
+// #include <string.h>
+// #include <stdio.h>
+//
+// int ra_tls_create_key_and_crt_der_wrapper(void* f, u_int8_t** der_key, size_t* der_key_size, u_int8_t** der_crt, size_t* der_crt_size) {
+//      int ( * ra_tls_create_key_and_crt_der)(u_int8_t**, size_t*, u_int8_t**, size_t*);
+//      ra_tls_create_key_and_crt_der = (int (*)(u_int8_t**, size_t*, u_int8_t**, size_t*)) f;
+//      return ra_tls_create_key_and_crt_der(der_key, der_key_size, der_crt, der_crt_size);
+// }
+import "C"
+
+import (
+	"os"
+	"unsafe"
+)
+
+func getAttestationType() (string, error) {
+	attestationType, err := os.ReadFile("/dev/attestation/attestation_type")
+	if err != nil {
+		PrintDebug("SGX RA-TLS attestation type file '/dev/attestation/attestation_type' not found")
+		return "", RATLS_WRAPPER_ERR_LIB_LOAD_FAILED
+	}
+
+	return string(attestationType), nil
+}
+
+func RATLSCreateKeyAndCrtDer(crtPath string, keyPath string) error {
+	if ra_tls_attest_create_key_and_crt_der_callback_f == nil {
+		PrintDebug("RA-TLS attest libraries not linked.")
+		return RATLS_WRAPPER_ERR_LIB_LOAD_FAILED
+	}
+
+	attestationType, err := getAttestationType()
+	if err != nil {
+		return err
+	}
+
+	switch attestationType {
+	case "none":
+		PrintDebug("Skipping certificate creation. Remote attestation type: ", attestationType)
+		return RATLS_WRAPPER_CERTIFICATE_CREATION_FAILED
+	case "dcap":
+		var derCrt *C.uchar
+		var crtLen C.size_t
+		var derKey *C.uchar
+		var keyLen C.size_t
+
+		ret := C.ra_tls_create_key_and_crt_der_wrapper(ra_tls_attest_create_key_and_crt_der_callback_f, &derKey, &keyLen, &derCrt, &crtLen)
+
+		if ret != 0 {
+			PrintDebug("RATLSCreateKeyAndCrtDer failed with error ", ret)
+			// TODO: custom error type for ret
+			return ErrorCode(ret)
+		}
+
+		f, err := os.Create(crtPath)
```

Comment:
I'm not sure about writing the key and cert to disk. After all - it is highly confidential information, so there might be use cases, where the developer would like to only keep them in memory ( just like we do BTW, but we use tmpfs ). What do you think?

Reply:
The alternative would be to return only der cert and key. This would also give the application the possibility to convert them to PEM before writing them to disk.

Reply:
Thanks, nice suggestion. I'm inclined towards returning the cert and key and giving the application the liberty to use it directly from memory or save it, as saving and loading again doesn't make sense.